game-core/src/test/java/games/strategy/triplea/ai/pro/util/MoveBatcherTest.java

```diff
@@ -0,0 +1,130 @@
+    moves.addMove(unitList(transport2, inf2, tank2), sz18ToSz12);
+    moves.addMove(unitList(inf2, tank2), sz12ToAlgeria);
+
+    final List<Collection<Unit>> moveUnits = new ArrayList<>();
+    final List<Route> moveRoutes = new ArrayList<>();
+    final List<Map<Unit, Unit>> unitsToTransports = new ArrayList<>();
+    moves.batchAndEmit(moveUnits, moveRoutes, unitsToTransports);
+
+    assertEquals(3, moveUnits.size());
+    assertEquals(3, moveRoutes.size());
+    assertEquals(3, unitsToTransports.size());
+
+    assertEquals(unitList(inf1, tank1, tank2, inf2), moveUnits.get(0));
+    assertEquals(brazilToSz18, moveRoutes.get(0));
+    final var expected =
+        Map.of(inf1, transport1, tank1, transport1, inf2, transport2, tank2, transport2);
+    assertTrue(Maps.difference(expected, unitsToTransports.get(0)).areEqual());
+
+    assertEquals(unitList(transport1, tank1, inf1, transport2, inf2, tank2), moveUnits.get(1));
+    assertEquals(sz18ToSz12, moveRoutes.get(1));
+    assertEquals(null, unitsToTransports.get(1));
+
+    assertEquals(unitList(tank1, inf1, inf2, tank2), moveUnits.get(2));
+    assertEquals(sz12ToAlgeria, moveRoutes.get(2));
+    assertEquals(null, unitsToTransports.get(2));
+  }
+
+  @Test
+  public void testTransportsPickingUpUnitsOnTheWay() {
+    final MoveBatcher moves = new MoveBatcher();
+
+    moves.newSequence();
+    moves.addMove(unitList(transport1), sz19ToSz18);
+    moves.addTransportLoad(tank1, brazilToSz18, transport1);
+    moves.addMove(unitList(transport1, tank1), sz18ToSz12);
+    moves.addMove(unitList(tank1), sz12ToAlgeria);
+    moves.newSequence();
+    moves.addMove(unitList(transport2), sz19ToSz18);
+    moves.addTransportLoad(inf1, brazilToSz18, transport2);
+    moves.addTransportLoad(inf2, brazilToSz18, transport2);
+    moves.addMove(unitList(transport2, inf1, inf2), sz18ToSz12);
+    moves.addMove(unitList(inf1, inf2), sz12ToAlgeria);
+
+    final List<Collection<Unit>> moveUnits = new ArrayList<>();
+    final List<Route> moveRoutes = new ArrayList<>();
+    final List<Map<Unit, Unit>> unitsToTransports = new ArrayList<>();
+    moves.batchAndEmit(moveUnits, moveRoutes, unitsToTransports);
+
+    assertEquals(4, moveUnits.size());
+    assertEquals(4, moveRoutes.size());
+    assertEquals(4, unitsToTransports.size());
+
+    assertEquals(unitList(transport1, transport2), moveUnits.get(0));
+    assertEquals(sz19ToSz18, moveRoutes.get(0));
+    assertEquals(null, unitsToTransports.get(0));
+
+    assertEquals(unitList(tank1, inf1, inf2), moveUnits.get(1));
+    assertEquals(brazilToSz18, moveRoutes.get(1));
+    final var expected = Map.of(tank1, transport1, inf1, transport2, inf2, transport2);
+    assertTrue(Maps.difference(expected, unitsToTransports.get(1)).areEqual());
+
+    assertEquals(unitList(transport1, tank1, transport2, inf1, inf2), moveUnits.get(2));
+    assertEquals(sz18ToSz12, moveRoutes.get(2));
+    assertEquals(null, unitsToTransports.get(2));
```

Comment:
Out of curiosity, why are null elements stored? Wouldn't it be just as well to not even store the element? EG: is an index out of bounds perhaps expected here?

Reply:
It's the same messy parallel list business.

unitsToTransports.get(2) corresponds to moveRoutes.get(2) to moveUnits.get(2).